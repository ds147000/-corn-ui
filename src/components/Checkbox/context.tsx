import React from 'react'

export type CheckBoxContextValue = {
  values?: string[]
  onCheck?(check: boolean, value: string): void
}

const DDEFAULT_CHECKBOX_CONTEXT_VALUE: CheckBoxContextValue = {}

export const CheckBoxContext = React.createContext(DDEFAULT_CHECKBOX_CONTEXT_VALUE)
