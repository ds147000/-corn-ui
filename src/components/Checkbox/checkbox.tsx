import React, { useContext, useEffect, useState } from 'react'
import classNames from 'classnames'
import {
  View,
  // #if _APP === 'weapp'
  Input
  // #endif
} from '@tarojs/components'
import Icon from '../Icon'
import Button, { ButtonSize, ButtonType } from '../Button'
import { CheckBoxContext } from './context'
import { FromContext } from '../Form/context'

export interface CheckBoxProps {
  name?: string
  /** 复选框类型 */
  type?: 'button' | 'default'
  /** 是否选中 */
  check?: boolean
  /** 默认是否选中 */
  defaultChecked?: boolean
  /** 值，仅在 checkGroup 嵌套下生效 */
  value?: string
  /** 禁用 */
  disabled?: boolean
  /** 大小，当类型为Button生效 */
  size?: ButtonSize
  /** 是否开启幽灵样式，仅在button有效 */
  ghost?: boolean
  /** 按钮风格样式 */
  buttonType?: ButtonType
  /** 选中发生改变回调 */
  onChange?(check: boolean): void
}

const CheckBox: React.FC<CheckBoxProps> = ({
  children,
  defaultChecked = false,
  check,
  type = 'default',
  size,
  onChange,
  value = '',
  disabled,
  name,
  ghost,
  buttonType = 'primary'
}) => {
  const [ internalCheck, setInternalCheck ] = useState(defaultChecked || false)
  const context = useContext(CheckBoxContext)
  const formContext = useContext(FromContext)

  /** 选中值 */
  const controllCheck = getControllValue(check, internalCheck, context.values, value)
  /** 是否默认样式 */
  const isDefualtStyle = type === 'default'
  /** 外部是否存在包裹组 */
  const isGroup = Boolean(context.onCheck)
  /** 表单值 */
  const _value = value || JSON.stringify(controllCheck)
  /** 控制表单名称 */
  const _name  = isGroup ? '' : name

  const _class = classNames(
    'xrk-if xrk-ac',
    controllCheck && isDefualtStyle && 'xrk-checkbox-active',
    isDefualtStyle && 'xrk-checkbox',
    isDefualtStyle && disabled && 'xrk-checkbox-disable'
  )

  let input: JSX.Element | null = null
  // #if _APP === 'weapp'
  input = <Input className="xrk-checkbox-hide" value={_value} name={_name} data-testid="check" />
  // #else
  input = (
  // eslint-disable-next-line react/forbid-elements
    <input
      className="xrk-checkbox-hide"
      value={_value}
      name={_name}
      data-testid="check"
      readOnly
    />
  )
  // #endif

  const _onChange = (): void => {
    if (disabled) return

    const newCheck = !controllCheck
    onChange?.(newCheck)
    context.onCheck?.(newCheck, value)
    if (check === undefined && context.values === undefined) setInternalCheck(newCheck)
  }

  useEffect(() => {
    if (!_name) return undefined

    formContext.add(_name, (val: unknown): void => setInternalCheck(Boolean(val)))

    return (): void => formContext.remove(_name)
  }, [ formContext, _name ])

  return (
    <View className={_class} onClick={_onChange} data-testid="xrk-checkbox" >
      {input}
      {isDefualtStyle ? (
        <>
          <Icon name={controllCheck ? 'fill-select' : 'radio'} />
          {children}
        </>
      ) : (
        <Button
          size={size}
          type={controllCheck ? buttonType : 'default'}
          disabled={disabled}
          ghost={controllCheck ? ghost : false}
        >
          {children}
        </Button>
      )}
    </View>
  )
}

export function getControllValue(
  externalCheck: boolean | undefined,
  internalCheck: boolean,
  values?: string[],
  value = ''
): boolean {
  if (values !== undefined) return values.includes(value)

  return externalCheck ?? internalCheck
}

export default CheckBox
