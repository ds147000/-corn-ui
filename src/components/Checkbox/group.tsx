import React from 'react'
import { View } from '@tarojs/components'
// #if _APP === 'weapp'
import Taro from '@tarojs/taro'
// #endif
import { CheckBoxContext, CheckBoxContextValue } from './context'
import { ViewProps } from '../../types/View'

export interface CheckBoxGroupProps extends ViewProps {
  name?: string
  defaultValue?: string[]
  onChange?(value: string[]): void
  radio?: boolean
}

class CheckBoxGroup extends React.Component<CheckBoxGroupProps, CheckBoxContextValue> {
  state: CheckBoxContextValue = {
    values: []
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ref: any

  private onChange = (check: boolean, value: string): void => {
    if (!this.props.radio) {
      this.onCheckBoxSelect(check, value)
    } else {
      this.onRadioSelect(check, value)
    }
  }

  /** 多选状态选中 */
  private onCheckBoxSelect = (check: boolean, value: string): void => {
    let newValues

    if (check) {
      newValues = [ ...this.state.values as string[], value ]
    } else {
      newValues = this.state.values?.filter((item) => item !== value)
    }

    this.setState({ values: newValues }, this.emitChange)
  }

  /** 单选状态选中 */
  private onRadioSelect = (check: boolean, value: string): void => {
    if (!check) return // 不可取消选中

    this.setState({ values: [ value ] }, this.emitChange)
  }

  /** 通知外部更新 */
  private emitChange = (): void => {
    this.props.onChange?.(this.state.values as string[])
  }

  // #if _APP === 'h5'
  private selectAllOfH5(): void {
    const ref: HTMLDivElement = this.ref
    const elements = ref.getElementsByTagName('input')
    const values: string[] = []

    for(let i = 0; i < elements.length; i++) {

      if (values.includes(elements[i].value) === false)
        values.push(elements[i].value)
    }

    this.setState({ values }, this.emitChange)
  }
  // #endif

  // #if _APP === 'weapp'
  private selectAllOfWeapp(): void {
    Taro.createSelectorQuery()
      .selectAll(`#${this.ref.uid} .xrk-checkbox-hide`)
      .fields({ properties: [ 'value' ] }, (list): void => {
        const values: string[] = list?.map((item) => item.value)
        this.setState({ values }, this.emitChange)
      }).exec()
  }
  // #endif

  /** 重置所有值 */
  reset = (): void => {
    this.setState({ values: [] })
  }

  /**
   * 设置表单值
   * @param string
   */
  setValue = (val: string[]): void => {
    this.setState({ values: [ ...val ] })
  }

  /** 全选 */
  selectAll = (): void => {
    if (this.props.radio) return // 单选，终止

    // #if _APP === 'weapp'
    this.selectAllOfWeapp()
    // #else
    this.selectAllOfH5()
    // #endif
  }

  /** 获取表单值 */
  getValue(): string[] {
    return [ ...this.state.values as string[] ]
  }

  componentDidMount(): void {
    if (this.props.defaultValue) this.setState({ values: [ ...this.props.defaultValue ] })
  }

  render(): JSX.Element {
    const ProviderValue: CheckBoxContextValue = {
      values: this.state.values,
      onCheck: this.onChange
    }

    const { children, ...props } = this.props

    return (
      <View ref={(ref): void => this.ref = ref} {...props} >
        <CheckBoxContext.Provider value={ProviderValue}>
          {children}
        </CheckBoxContext.Provider>
      </View>
    )
  }
}

export default CheckBoxGroup
