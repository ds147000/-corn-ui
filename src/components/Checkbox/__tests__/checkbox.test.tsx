import React, { useRef, useState } from 'react'
import Checkbox, { getControllValue } from '../checkbox'
import { fireEvent, render, waitFor } from '@testing-library/react'

describe('Checkbox', () => {
  test('基础', () => {
    const screen = render(<Checkbox>商品链接</Checkbox>)
    screen.getByText('商品链接')
    expect(screen.container).toMatchSnapshot()
  })

  test('基础 按钮类型', () => {
    const screen = render(<Checkbox type="button" >商品链接</Checkbox>)
    screen.getByText('商品链接')
    expect(screen.container).toMatchSnapshot()
  })

  test('基础 按钮类型', () => {
    const screen = render(<Checkbox type="button" check >商品链接</Checkbox>)
    screen.getByText('商品链接')
    expect(screen.container).toMatchSnapshot()
  })


  test('选中', () => {
    const screen = render(<Checkbox check >商品链接</Checkbox>)
    screen.getByText('商品链接')
    screen.getByTestId('check')
    expect(screen.container).toMatchSnapshot()
  })

  test('值', () => {
    const screen = render(<Checkbox value="100" >商品链接</Checkbox>)
    screen.getByText('商品链接')
  })

  test('操作', () => {
    const screen = render(<Checkbox>商品链接</Checkbox>)
    fireEvent.click(screen.getByText('商品链接'))
    expect(screen.getByTestId('check'))
  })

  test('操作2', () => {
    const onChange = jest.fn()
    const screen = render(<Checkbox onChange={onChange} >商品链接</Checkbox>)
    fireEvent.click(screen.getByText('商品链接'))
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange.mock.calls[0][0]).toBe(true)
  })

  test('操作3', () => {
    const onChange = jest.fn()
    const screen = render(<Checkbox onChange={onChange} defaultChecked={true} >商品链接</Checkbox>)
    fireEvent.click(screen.getByText('商品链接'))
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange.mock.calls[0][0]).toBe(false)
  })

  test('操作 反复', () => {
    const onChange = jest.fn()
    const screen = render(<Checkbox onChange={onChange} defaultChecked={true} >商品链接</Checkbox>)
    fireEvent.click(screen.getByText('商品链接'))
    fireEvent.click(screen.getByText('商品链接'))
    expect(onChange).toHaveBeenCalledTimes(2)
    expect(onChange.mock.calls[0][0]).toBe(false)
    expect(onChange.mock.calls[1][0]).toBe(true)
  })

  test('禁用操作', () => {
    const onChange = jest.fn()
    const screen = render(<Checkbox onChange={onChange} disabled >商品链接</Checkbox>)
    fireEvent.click(screen.getByText('商品链接'))
    expect(onChange).toHaveBeenCalledTimes(0)
  })

  test('自控制', async () => {
    const App: React.FC = () => {
      const [check, setCheck] = useState(false)
      const num = useRef(0)

      const onChange = () => {
        num.current += 1
        if (num.current % 2 === 0) setCheck(!check)
      }
      return (<Checkbox check={check} onChange={onChange} >商品链接</Checkbox>)
    }

    const screen = render(<App />)
    fireEvent.click(screen.getByText('商品链接'))
    await waitFor(() => expect(screen.getByTestId('check').getAttribute('value')).toBe('false'))
    fireEvent.click(screen.getByText('商品链接'))
    await waitFor(() => expect(screen.getByTestId('check').getAttribute('value')).toBe('true'))
    fireEvent.click(screen.getByText('商品链接'))
    await waitFor(() => expect(screen.getByTestId('check').getAttribute('value')).toBe('true'))
  })
})

describe('Checkbox/getControllValue', () => {
  test('内部变量控制', () => {
    expect(getControllValue(undefined, false)).toBe(false)
    expect(getControllValue(undefined, true)).toBe(true)
  })

  test('外部变量控制', () => {
    expect(getControllValue(true, false)).toBe(true)
    expect(getControllValue(false, true)).toBe(false)
  })

  test('Context控制', () => {
    expect(getControllValue(true, false, [])).toBe(false)
    expect(getControllValue(false, true, ['1'])).toBe(false)
    expect(getControllValue(false, true, ['1'], '2')).toBe(false)
    expect(getControllValue(false, true, [], '2')).toBe(false)
    expect(getControllValue(false, true, ['1', '2'], '2')).toBe(true)
    expect(getControllValue(false, true, ['3'], '3')).toBe(true)
  })
})
