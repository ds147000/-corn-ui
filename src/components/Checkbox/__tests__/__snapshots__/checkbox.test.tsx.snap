// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox 基础 1`] = `
<div>
  <div
    class="xrk-if xrk-ac xrk-checkbox"
  >
    <input
      class="xrk-checkbox-hide"
      data-testid="check"
      readonly=""
      value="false"
    />
    <i
      class="xrk-icon xrk-i-radio"
    />
    商品链接
  </div>
</div>
`;

exports[`Checkbox 基础 按钮类型 1`] = `
<div>
  <div
    class="xrk-if xrk-ac"
  >
    <input
      class="xrk-checkbox-hide"
      data-testid="check"
      readonly=""
      value="false"
    />
    <div
      class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-default"
    >
      商品链接
    </div>
  </div>
</div>
`;

exports[`Checkbox 基础 按钮类型 2`] = `
<div>
  <div
    class="xrk-if xrk-ac"
  >
    <input
      class="xrk-checkbox-hide"
      data-testid="check"
      readonly=""
      value="true"
    />
    <div
      class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-primary"
    >
      商品链接
    </div>
  </div>
</div>
`;

exports[`Checkbox 选中 1`] = `
<div>
  <div
    class="xrk-if xrk-ac xrk-checkbox-active xrk-checkbox"
  >
    <input
      class="xrk-checkbox-hide"
      data-testid="check"
      readonly=""
      value="true"
    />
    <i
      class="xrk-icon xrk-i-fill-select"
    />
    商品链接
  </div>
</div>
`;
